BraintreeCore/src/test/java/com/braintreepayments/api/AuthorizationLoaderUnitTest.java

```diff
@@ -0,0 +1,151 @@
+        verify(callback).onAuthorizationResult(captor.capture(), (Exception) isNull());
+
+        Authorization authorization = captor.getValue();
+        assertEquals(initialAuthString, authorization.toString());
+    }
+
+    @Test
+    public void loadAuthorization_whenInitialAuthDoesNotExist_callsBackSuccessfulClientTokenFetch() {
+        String clientToken = Fixtures.BASE64_CLIENT_TOKEN;
+        ClientTokenProvider clientTokenProvider = new MockClientTokenProviderBuilder()
+                .clientToken(clientToken)
+                .build();
+        sut = new AuthorizationLoader(null, clientTokenProvider);
+
+        AuthorizationCallback callback = mock(AuthorizationCallback.class);
+        sut.loadAuthorization(callback);
+
+        ArgumentCaptor<Authorization> captor = ArgumentCaptor.forClass(Authorization.class);
+        verify(callback).onAuthorizationResult(captor.capture(), (Exception) isNull());
+
+        Authorization authorization = captor.getValue();
+        assertEquals(clientToken, authorization.toString());
+    }
+
+    @Test
+    public void loadAuthorization_whenInitialAuthDoesNotExist_cachesClientTokenInMemory() {
+        String clientToken = Fixtures.BASE64_CLIENT_TOKEN;
+        ClientTokenProvider clientTokenProvider = new MockClientTokenProviderBuilder()
+                .clientToken(clientToken)
+                .build();
+        sut = new AuthorizationLoader(null, clientTokenProvider);
+
+        AuthorizationCallback callback = mock(AuthorizationCallback.class);
+        sut.loadAuthorization(callback);
+        sut.loadAuthorization(callback);
+
+        verify(clientTokenProvider, times(1)).getClientToken(any(ClientTokenCallback.class));
+    }
+
+    @Test
+    public void loadAuthorization_whenInitialAuthDoesNotExist_forwardsClientTokenFetchError() {
+        Exception clientTokenFetchError = new Exception("error");
+        ClientTokenProvider clientTokenProvider = new MockClientTokenProviderBuilder()
+                .error(clientTokenFetchError)
+                .build();
+        sut = new AuthorizationLoader(null, clientTokenProvider);
+
+        AuthorizationCallback callback = mock(AuthorizationCallback.class);
+        sut.loadAuthorization(callback);
+
+        verify(callback).onAuthorizationResult(null, clientTokenFetchError);
+    }
+
+    @Test
+    public void loadAuthorization_whenInitialAuthDoesNotExistAndNoClientTokenProvider_callsBackException() {
+        sut = new AuthorizationLoader(null, null);
+
+        AuthorizationCallback callback = mock(AuthorizationCallback.class);
+        sut.loadAuthorization(callback);
+
+        ArgumentCaptor<Exception> captor = ArgumentCaptor.forClass(Exception.class);
+        verify(callback).onAuthorizationResult((Authorization)isNull(), captor.capture());
+
+        Exception error = captor.getValue();
+        assertTrue(error instanceof BraintreeException);
+        assertEquals("Unable to fetch client token", error.getMessage());
```

Comment:
Maybe this error message should be more specific in this case? Should we have the "Unable to fetch" as the default and return a specific error message about requiring authorization string or auth provider if both params are null?

Reply:
Yeah definitely should be improved. I added a dev docs URL. This happens when `null` is passed in for either an auth string or a `ClientTokenProvider`. Our API says `@NonNull` for either parameter but it's still technically possible.